clock/README.md

```diff
@@ -5,6 +5,362 @@ A Go time testing library for writing deterministic unit tests
+})
+
+w := mClock.Advance(time.Second)
+err := w.Wait(ctx)
+if err != nil {
+  t.Fatal("AfterFunc f never completed")
+}
+if !fired {
+  t.Fatal("didn't fire")
+}
+```
+
+The construction of waiting for the triggered events and failing the test if they don't complete is
+very common, so there is a shorthand:
+
+```go
+w := mClock.Advance(time.Second)
+err := w.Wait(ctx)
+if err != nil {
+  t.Fatal("AfterFunc f never completed")
+}
+```
+
+is equivalent to:
+
+```go
+w := mClock.Advance(time.Second)
+w.MustWait(ctx)
+```
+
+or even more briefly:
+
+```go
+mClock.Advance(time.Second).MustWait(ctx)
+```
+
+### Advance only to the next event
+
+One important restriction on advancing the clock is that you may only advance forward to the next
+timer or ticker event and no further. The following will result in a test failure:
+
+```go
+func TestAdvanceTooFar(t *testing.T) {
+    ctx, cancel := context.WithTimeout(10*time.Second)
+    defer cancel()
+    mClock := quartz.NewMock(t)
+    var firedAt time.Time
+    mClock.AfterFunc(time.Second, func() {
+        firedAt := mClock.Now()
+    })
+    mClock.Advance(2*time.Second).MustWait(ctx)
+}
+```
+
+This is a deliberate design decision to allow `Advance()` to immediately and synchronously move the
+clock forward (even without calling `Wait()` on returned waiter). This helps meet Quartz's design
+goals of writing deterministic and easy to understand unit tests. It also allows the clock to be
+advanced, deterministically _during_ the execution of a tick or timer function, as explained in the
+next sections on Traps.
+
+Advancing multiple events can be accomplished via looping. E.g. if you have a 1-second ticker
+
+```go
+for i:=0; i<10; i++ {
+    mClock.Advance(time.Second).MustWait(ctx)
+}
+```
+
+will advance 10 ticks.
+
+If you don't know or don't want to compute the time to the next event, you can use `AdvanceNext()`.
+
+```go
+d, w := mClock.AdvanceNext()
+w.MustWait(ctx)
+// d contains the duration we advanced
+```
+
+### Traps
+
+A trap allows you to match specific calls into the library while mocking, block their return,
+inspect their arguments, then release them to allow them to return. They help you write
+deterministic unit tests even when the code under test executes asynchronously from the test.
+
+You set your traps prior to executing code under test, and then wait for them to be triggered.
+
+```go
+func TestTrap(t *testing.T) {
+    ctx, cancel := context.WithTimeout(10*time.Second)
+    defer cancel()
+    mClock := quartz.NewMock(t)
+    trap := mClock.Trap().AfterFunc()
+    defer trap.Close() // stop trapping AfterFunc calls
+
+    count := 0
+    go mClock.AfterFunc(time.Hour, func(){
+        count++
+    })
+    call := trap.MustWait(ctx)
+    call.Release()
+    if call.Duration != time.Hour {
+        t.Fatal("wrong duration")
+    }
+
+    // Now that the async call to AfterFunc has occurred, we can advance the clock to trigger it
+    mClock.Advance(time.Hour).MustWait(ctx)
```

Comment:
```suggestion
    mClock.Advance(call.Duration).MustWait(ctx)
```

Not sure if this kind of usage should be encouraged or discouraged, thoughts?

Reply:
Not sure it matters one way or another in this example, since we assert that these are equal just above.

This would be fine if the intent of the test _doesn't_ depend on the actual length of the timer. In an abstract example like this, who can say...